src/builtin_cd.cpp

```diff
@@ -47,55 +47,69 @@ maybe_t<int> builtin_cd(parser_t &parser, io_streams_t &streams, wchar_t **argv)
     }
 
     wcstring pwd = parser.vars().get_pwd_slash();
-    maybe_t<wcstring> mdir = path_get_cdpath(dir_in, pwd, parser.vars());
-    if (!mdir) {
-        if (errno == ENOTDIR) {
-            streams.err.append_format(_(L"%ls: '%ls' is not a directory\n"), cmd, dir_in.c_str());
-        } else if (errno == ENOENT) {
-            streams.err.append_format(_(L"%ls: The directory '%ls' does not exist\n"), cmd,
-                                      dir_in.c_str());
-        } else if (errno == EROTTEN) {
-            streams.err.append_format(_(L"%ls: '%ls' is a rotten symlink\n"), cmd, dir_in.c_str());
-        } else {
-            streams.err.append_format(_(L"%ls: Unknown error trying to locate directory '%ls'\n"),
-                                      cmd, dir_in.c_str());
-        }
+    auto dirs = path_apply_cdpath(dir_in, pwd, parser.vars());
+    if (dirs.empty()) {
+        streams.err.append_format(_(L"%ls: The directory '%ls' does not exist\n"), cmd,
+                                  dir_in.c_str());
 
         if (!parser.is_interactive()) streams.err.append(parser.current_line());
 
         return STATUS_CMD_ERROR;
     }
-    const wcstring &dir = *mdir;
-
-    wcstring norm_dir = normalize_path(dir);
-
-    // We need to keep around the fd for this directory, in the parser.
-    autoclose_fd_t dir_fd(wopen_cloexec(norm_dir, O_RDONLY));
-    bool success = dir_fd.valid() && fchdir(dir_fd.fd()) == 0;
 
-    if (!success) {
-        struct stat buffer;
-        int status;
-
-        status = wstat(dir, &buffer);
-        if (!status && S_ISDIR(buffer.st_mode)) {
-            streams.err.append_format(_(L"%ls: Permission denied: '%ls'\n"), cmd, dir_in.c_str());
-        } else {
-            streams.err.append_format(_(L"%ls: '%ls' is not a directory\n"), cmd, dir_in.c_str());
+    errno = 0;
+    auto best_errno = errno;
+
+    for (auto &dir: dirs) {
+        wcstring norm_dir = normalize_path(dir);
+
+        // We need to keep around the fd for this directory, in the parser.
+        errno = 0;
+        autoclose_fd_t dir_fd(wopen_cloexec(norm_dir, O_RDONLY));
+        bool success = dir_fd.valid() && fchdir(dir_fd.fd()) == 0;
+
+        if (!success) {
+            // If the directory doesn't exist, we ignore it and only error if none does.
+            // ENOENT is a low-priority error - if we found something with the name that isn't a directory, we error out.
```

Comment:
This refers to the ENOTDIR case, right? The ENOENT case is described in the line above.
Also the second sentence is a bit confusing because we don't error out immediately but still try other CDPATH entries but I don't know how to word it differently.